### PR TITLE
Opt fix

### PR DESCRIPTION
**Pull request by nouiz**
fix gh-4220

### CODE REVIEW

File: theano/sandbox/cuda/opt.py [hunk 1]
```diff
@@ -1942,7 +1942,9 @@ def local_gpu_downsample_factor_max_grad_grad(node):
     if isinstance(node.op, pool.DownsampleFactorMaxGradGrad):
         assert node.op.__props__ == ('ds', 'ignore_border', 'st',
                                      'padding', 'mode')
-        if node.op.padding != (0, 0) or node.op.mode != 'max':
+        if (node.op.padding != (0, 0) or
+                node.op.mode != 'max' or
+                node.op.st != (1, 1)):
```
Comment (f0k) [inline theano/sandbox/cuda/opt.py]: :+1:

File: theano/sandbox/cuda/dnn.py [hunk 2]
```diff
@@ -2246,9 +2247,13 @@ def local_dnn_conv_inplace(node):
         inputs = list(node.inputs)
         dest = inputs[2]
         if (dest.owner and
-                isinstance(dest.owner.op, GpuAllocEmpty) and
+                type(dest.owner.op) is GpuAllocEmpty and
```
Comment (f0k) [inline theano/sandbox/cuda/dnn.py]: What exactly is the difference?

Comment (lamblin) [inline theano/sandbox/cuda/dnn.py]: `GpuAlloc` inherits from `GpuAllocEmpty`, so the previous check would have matched `GpuAlloc` nodes as well.

**Comment by lamblin** [comment]
Are those fixes needed on the new back-end as well?

File: theano/sandbox/cuda/dnn.py (hunk 2 above)
Comment (f0k) [inline theano/sandbox/cuda/dnn.py]: > `GpuAlloc` inherits from `GpuAllocEmpty`

Oh, I forgot. Wouldn't it be better to just swap the cases then, so we can still use `isinstance`?
```python
if ... and isinstance(..., GpuAlloc) and ...:
    ...
elif ... and isinstance(..., GpuAllocEmpty) and ...:
    ...
```
Directly checking for the type could be problematic if additional subclasses of GpuAlloc or GpuAllocEmpty are introduced.

**Comment by nouiz** [comment]
They aren't needed in the new back-end as they don't inherit from each other.

File: theano/sandbox/cuda/dnn.py (hunk 2 above)
Comment (nouiz) [inline theano/sandbox/cuda/dnn.py]: In the new back-end, we use type equality. We don't want subclass to get optimized by default. We prefer them to tell explicitly they want to be optimized. This is safer (we prefer to risk less optimized graph then bugged graph). Doing that here isn't a problem I think.

File: theano/sandbox/cuda/opt.py (hunk 1 above)
Comment (lamblin) [inline theano/sandbox/cuda/opt.py]: Actually, I think this should be `node.op.st != node.op.ds`.
A fix is on the way once I check that it's correct.

Comment (lamblin) [inline theano/sandbox/cuda/opt.py]: See #4249.

**Pull request merged**